crazy_rl/multi_agent/jax/catch/catch.py

```diff
@@ -0,0 +1,273 @@
+            low=np.tile(np.array([-self.size, -self.size, 0], dtype=np.float32), self.num_drones + 1),
+            high=np.tile(np.array([self.size, self.size, self.size], dtype=np.float32), self.num_drones + 1),
+            shape=(3 * (self.num_drones + 1),),  # coordinates of the drones and the target
+            dtype=np.float32,
+        )
+
+    @override
+    def _action_space(self):
+        return spaces.Box(low=-1 * np.ones(3, dtype=np.float32), high=np.ones(3, dtype=np.float32), dtype=np.float32)
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_obs(self, state, key):
+        key, subkey = random.split(key)
+
+        # mean of the agent's positions
+        mean = state.agents_locations.sum() / self.num_drones
+
+        dist = jnp.linalg.norm(mean - state.target_location)
+
+        surrounded = dist <= 0.2
+
+        # if the target is out of the map, put it back in the map
+        target_location = jnp.clip(
+            state.target_location
+            + (
+                # go to the opposite direction of the mean of the agents
+                (1 - surrounded)
+                * (state.target_location - mean)
+                / dist
+                * self.target_speed
+            )
+            + (
+                # if the mean of the agents is too close to the target, move the target in a random
+                # direction, slowly because it hesitates
+                surrounded
+                * random.uniform(subkey, (3,), minval=-1, maxval=1)
+                * self.target_speed
+                * 0.1
+            ),
+            jnp.array([-self.size, -self.size, 0.2]),
+            jnp.array([self.size, self.size, self.size]),
+        )
+
+        return (
+            jdc.replace(
+                state,
+                observations=jnp.append(
+                    jnp.column_stack((state.agents_locations, jnp.tile(target_location, (self.num_drones, 1)))),
+                    jnp.array(
+                        [jnp.delete(state.agents_locations, agent, axis=0).flatten() for agent in range(self.num_drones)]
+                    ),
+                    axis=1,
+                ),
+                target_location=target_location,
+            ),
+            key,
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_action(self, state, actions):
+        # Actions are clipped to stay in the map and scaled to do max 20cm in one step
+        return jdc.replace(
+            state,
+            agents_locations=jnp.clip(
+                state.agents_locations + actions * 0.2, jnp.array([-self.size, -self.size, 0]), self.size
+            ),
+        )
+
+    @override
+    @partial(jit, static_argnums=(0,))
+    def _compute_reward(self, state):
+        # Reward is the mean distance to the other agents plus a maximum value minus the distance to the target
+
+        return jdc.replace(
+            state,
+            rewards=jnp.any(state.truncations)
```

Comment:
you mean `1 - jnp.any(state.terminated)` ?

Reply:
It was in the case where we give the positive reward only at the end of the game.